### DISCUSSION TITLE
Progress Report

### DISCUSSION THREAD

**Discussion by CalvinJDP** [glazet/QuadEVCharger#18] [General]
<html>
<body>
<google-sheets-html-origin>

Date | Description
-- | --
10/16/23 - 10/22/23 | Purchased Single Board Computer Raspberry Pi 4 and Analog Digital Converter board, Wrote Skeleton Code for Algorithm, Created Overview, Hardware, and Software Block Diagram, Alternate Design Matrices

</body>
</html>

**Comment by glazet** [comment]
10/16/23 - 10/22/23
Purchased MCP3008 and TL494/TL594
Determined that the only user variables to consider are battery level and time parked. Time of day, vehicle size were also considered.
Decided to break charging into two stages. Stage One prioritizes getting every EV to commute charge. Stage Two is for charging EVs fully. Stage One likely will not consider user inputs too heavily while Stage Two consider user inputs much more
Created matrix equation based on battery level and time parked. The equation outputs a matrix of charge time in minutes, (for Stage Two)
Basic structure of code and interrupt was written and tested on RPi. (See code section)

**Comment by neilh10** [comment]
Thanks for the update.   Since your are primarily doing a "hardware verification", and not going to be able to ever verify this in operation - the "test fixture" of 4 vehicles and a 50A circuit is beyond your budget, it doesn't seem worth any effort in proposing anything other than a simple power/current sharing of the available power to vehicles via https://github.com/glazet/QuadEVCharger/issues/22.   
For the two algorithm's you are proposing - a) how do you know the EV battery state of charge? b)  how do you know what current each vehicle is taking?, the last diagram I saw only had one common current detector (pg9). Doesn't there need to be a current detector per EV circuit ?

**Reply by glazet** [reply]
a) how do you know the EV battery state of charge?
Answer: We are going to prompt the user to input their current battery percentage and their expected time of departure. 
b) how do you know what current each vehicle is taking? Doesn't there need to be a current detector per EV circuit ?
Answer: With this algorithm, we are only charging one EV at a time at 40A in a consecutive order in terms of first come first serve priority. This means we only need a single current transformer to keep track of current draw. 
We would know which car is being charged by what relay we have triggered. Do you mean how will we validate this?

**Comment by neilh10** [comment]
Great this is the value of writing it up. 
to be able to flush out the details you need. Not sure I follow your logic, and the person assigned to it needs to do the work to explain it.

Could you write up A and B as separate issues, in engineering language with the formulas you need to be able to determine the results. Then assign them to the team members who are doing the work.

So for example "Stage One prioritizes getting every EV to commute charge. Stage One likely will not consider user inputs too heavily while Stage"   - if I understand it "commute charge" is giving a charge to each vehicle, in measured Ahrs . In the write up ~ please specify the formulae's and Ahrs that perform that providing this engineering function.

The benefit of writing it up as an issue, is that some of these may be required to be reflected in the user interface,  other factors may come into play.  How does impact #20 "Metering", summing all the Ahrs for billing.?

**Comment by glazet** [comment]
10/22/23-10/30/23
Decided on a contactor for switching device. It is a Packard C240C, dual pole, can handle 208/240VAC and up to 50A continuous. It is activated with a 12V signal. 
Developed a rough draft of our tests, which can viewed from this [document](https://docs.google.com/document/d/1dsUsDy-1TQu-y7XCOsLgnBoxz6zdpGsGgdZrkOZMSAA/edit)
Developed a rough draft for our Gantt chart, but is still a work in progress. 
We have gone over our slides and report with Dr. Salem and have revised many parts of these documents. We got a lot of good feedback from him and got an understanding of what our minimum viable product would be. We should show our ability to create the control pilot signal, proximity pilot triggering, and switching device control based on EVSE connection status.

**Reply by neilh10** [reply]
Thanks for the update - can you provide a reference for Packard C240C 
Note for pp - https://github.com/glazet/QuadEVCharger/issues/21
For control pilot  https://github.com/glazet/QuadEVCharger/issues/22

**Comment by djangodemetri** [comment]
11/5/23
Accomplished:
1. Adjusted the [minimum viable product schedule](https://docs.google.com/spreadsheets/d/17dWSB3O24H2Vjkp8bGaF0v0rzq5ACB4X7yRKv1XJ_DY/edit#gid=550061121) (under Sheet 2 in the bottom left corner)

2. Simulated and adjusted the CP circuit for 3.3V

3. Simulated and built PP circuit

4. Attempted to get a ESP32-S3 DevKitC-1 to create PWM signal but was unable to get the S3 to flash new programs. It seems to be a deeper issue, certain people suggesting issues with Windows. Currently seeking help from forums and Expressif's help avenues. 
4.a) In the meantime, ordered ESP32S WROOM which has had PWM code successfully uploaded to it. This will likely be the SoC used for the minimum viable product but not for the final project. Why? Because the ESP32S does not have enough pins on the dev board to support our needs. So the DevKitC-1 will need to be worked on over winter break.

5. Ordered a [relay ](https://store.openevse.com/products/50a-dpst-relay-set) that should be here this week. (Datasheet is available as well). We only ordered one for proof of concept. (Due to its price we considered only having one for the final project as well).

Concerns:
1. The ESP32 should satisfy IEC 61851 tolerances but it if doesn't then we need to do a lot more work. The next step would be using a PIC18 that we are familiar with. It's overpowered for what the project needs but it would prove our ability for the minimum viable product.

2. The CP circuit that was found on OpenEVSE and brought to attention by Neil is unfamiliar to us. That implies we have many wrinkles to iron out and have some discrete electronics theory to refresh. 

To-do this week:
1. Validate the ESP32 can produce CP signal in tolerance

2. Create the CP circuit on a breadboard and began familiarizing ourselves with it

3. Measure the resistance of the ADC pin so that capacitor can be determined for the rectifier.

4. Now that we are transitioning to a microcontroller we have to adjust our report and slides. Sadness.

(Please note the 3 week look ahead is available in the [minimum viable product schedule](https://docs.google.com/spreadsheets/d/17dWSB3O24H2Vjkp8bGaF0v0rzq5ACB4X7yRKv1XJ_DY/edit#gid=550061121))

**Comment by djangodemetri** [comment]
11/13/23

Accomplishments:
1. ESP32 was able to create PWM with 5us tolerance (40-60ns were actually values) and 600mV tolerance (300mV actually value). 
2. Constructed and tested the [CP circuit (Slide 36)](https://www.fveaa.org/fb/J1772_386.pdf). It functioned as intended but was not able to produce a full 12V amplitude. Op-amp had maximum biasing voltage of 18V. Ordered a new op-amp to try again.
3. Reworked some slides in order to prepared for presentation

Concerns:
1. Largest concern is testing the plausibility of the current [engineering requirements](https://docs.google.com/presentation/d/1tK7BEZV_6pD6SzRpXh4eQvX8X0hkCamg/edit?usp=sharing&ouid=110144589343210258995&rtpof=true&sd=true) since they will be unchangeable after Thanksgiving. Unobtainable engineering requirements is an instant fail. 
2. Obtaining an op amp that satisfies the 600mV tolerance. According to this [Table A.1 (on page 41) and Figure A.1 (on page 40) of this manual](https://hmr.araiindia.com/Control/AIS/210201752832PMAIS138_Part1.pdf) the tolerances are only needed to be maintained at the Vg which is the generator. It is assumed that generator also encompasses the comparator. If this assumption is not true, then we do not have to satisfy the 600mV tolerance.

To-do this Week:
1. See concerns sections
2. Finalize slides. Meet with Sudhir, Salem, and Chris Stewart to get feedback on slides.

**Comment by djangodemetri** [comment]
Apologies for the lack of communication, it is now in the final section of the semester where everything accelerates quickly

Accomplishments:
1. The [EVSE-EV interface circuit (Slide 36)](https://www.fveaa.org/fb/J1772_386.pdf) has been constructed, tested, and proven to work correctly. The final circuit can be found on [Slide 19](https://docs.google.com/presentation/d/1tK7BEZV_6pD6SzRpXh4eQvX8X0hkCamg/edit?usp=sharing&ouid=110144589343210258995&rtpof=true&sd=true). As can be seen the LM331 is the current comparator in use, (subject to change), the powered resistor was removed and replaced by a passive rectifier, and a LM741 voltage follower was added as a buffer between the EV voltage divider and ESP32 voltage divider. The relay driver circuit from OpenEVSE was not effective so we pulled a simple driver circuit from our microprocessor class.

2. Trent used a calculator to determine any differences in charge effectiveness between dividing current and dividing time. Neither method appears to charge faster. This is likely predicted from viewing the simple voltage power equation P = E/t. If our power remains constant, (which it does), then however E scales, t must scale in opposition. For example, if you quarter the amperage, then you quadruple the time spent charging. The calculations can be found on the last [two slides](https://docs.google.com/presentation/d/1tK7BEZV_6pD6SzRpXh4eQvX8X0hkCamg/edit?usp=sharing&ouid=110144589343210258995&rtpof=true&sd=true).

3. We had the practice presentation and had good results. Please find the attached image to see the results.
![llll](https://github.com/glazet/QuadEVCharger/assets/131516746/adfa3773-d42e-431b-8aaf-da2600880d18)

4. Trent extended the problem statement to make clear that the EVSE is used for long term parking.

5. Calvin managed to get the ESP32-S3 functioning and ran a preliminary wifi test for ER 8. 
![lll](https://github.com/glazet/QuadEVCharger/assets/131516746/48c385e1-067e-483c-b392-1494f274cc14)

6. We decided to go with four current transformers. The final nail in the coffin was Salem's suggestion. "The customer wants to pay for the power they are receiving, not the power lost through all the components." We originally had a single current transformer on the mains. 

Concerns:

1. Time is now a premium 

2. Despite good reception from our practice presentation, burn out is knocking on our doors. (You can probably tell by how poorly this write up is worded ;) )

3. The presentation is only the highlights of the project. The report is where we will see all the holes in this project.

To-do this week:
1. We need to find the range for the CP voltages but have not been able to find anything from the technical documents we've found. Chris Stewart suggested we access IEC 61851 through the school's library and then we can find those voltage ranges.

2. FT1 and FT2, (satisfy +-5Hz and +-600mV), require a better understanding of the oscilloscope's precision to complete. We are meeting with Doctor Salem this evening to discuss this process.

3. Rework [engineering requirements (Slide 8)](https://docs.google.com/presentation/d/1tK7BEZV_6pD6SzRpXh4eQvX8X0hkCamg/edit?usp=sharing&ouid=110144589343210258995&rtpof=true&sd=true) YET AGAIN. We do not like ER8. In reference to ER1 through ER3, Chris Stewart said that we needed to mention the specific tolerances (+-5Hz and +-600mV) in the ER. They need to be reworded

4. Finish writing the report which is due on the 7th. 

Apologies again for the lack of communication.

**Reply by neilh10** [reply]
Good to see the report and results from @djangodemetri . The project is being flushed out. **Yay** and team members names are being referenced!!   

Great to see some of the detail and specifying slides - which gives me the ability to technically comment on it . Getting the right level of detail is key for comments.

Item 1]  I'm mystified why you are discussing an LM331 (Voltage to frequency converter) when you have the PWM function digital already specified from the ESP32-S3.  I thought you where on the right track with a standard comparator (741 style) but with Rail-to-rail output at 12V. I also suggested an alternative - clearly I wasted my time with the suggestion.   Whats up!  https://github.com/glazet/QuadEVCharger/issues/22#issuecomment-1809591316

Item 5
The ESP32 Family varies in devices - and therefore internal processors and functions. For a hardware engineer its critical to be clear about the devices.  The ESP32-S3 choice is good when discussing specifics.   Slide 21 the BOM specifies a ESP32-WROOM-2 which is a different device

Item 2] Analysis is too simplistic - what if you have a smart QuadEV charger, optimally sharing real-time charge with plugged in EV. The decision to go with 4 charging coils is right. Though 10A is mentioned elsewhere - how to apply full 40A

The choice of individual current monitoring to derive KWh delivered is critical for a billing interface.
I attach how my JuiceNet EV charger represents the charge provided to my EV,
Of course the QuadEV hardware unit requires a uniquie hardware ID, and customer ID for up to 4 different customers plugged in. 
In my case the billing is by separate PGE meter.  I'm not suggesting you need to do anything, other than be clear that KWh data can be produced and delivered with Unique Hardware ID and customer ID
 This brings up uniquie IDs- need one for the hardware unit (possibly WiFi ID?)  - and variable unique ID for billing, derived through NFC

[231202 JuiceNet Sessions.csv](https://github.com/glazet/QuadEVCharger/files/13537172/231202.JuiceNet.Sessions.csv)

and then this visually summarized as a monthly history. The graph doesn't state what the difference is between the bars - but could be different rate schedule.  I try and keep my vehicle on the Off Peak Time Of Use.

![231202 JuiceNet Past Year Monthly Energy History](https://github.com/glazet/QuadEVCharger/assets/2766619/60604f17-87eb-458c-a407-1ccfa4728c07)

Finally the Concerns identified for this stage of the project are not technical.   I don't see any movement of flushing out technical issues in https://github.com/glazet/QuadEVCharger/issues
